src/mlpack/core/data/detect_file_type_impl.hpp

```diff
@@ -286,12 +334,80 @@ void DetectFromExtension(const std::string& filename,
   {
     opts.Format() = FileType::ARFFASCII;
   }
+  else if constexpr (HasSerialize<ObjectType>::value)
+  {
+    if (extension == "xml")
+    {
+      opts.Format() = FileType::XML;
+    }
+    else if (extension == "bin")
+    {
+      opts.Format() = FileType::BIN;
+    }
+    else if (extension == "json")
+    {
+      opts.Format() = FileType::JSON;
+    }
+  }
```

Comment:
One issue here is that if the user passes a serializable type, but the filename is `.arff`, then what is returned is `FileType::ARFFASCII`.  I would say that either:

 * we should be sure that the returned type of `DetectFromExtension` is valid for the type of object, so it does not need to be checked outside this function, or
 * we should split `DetectFromExtension` into two versions, one for matrix types and one for serializable types.

Either situation means that the code in `DetectFileType` that checks that the format is valid can be significantly simplified.

It would be very easy when extending this code for more types later to end up with unexpected routes through `DetectFromExtension` that cause weird behavior when users give strange options.